File: fraiseql-php/src/Attributes/GraphQLType.php
<?php

declare(strict_types=1);

namespace FraiseQL\Attributes;

use Attribute;

/**
 * PHP 8 Attribute for defining GraphQL type definitions.
 *
 * Usage:
 * ```php
 * #[GraphQLType(name: 'User')]
 * class User {
 *     #[GraphQLField]
 *     public int $id;
 *
 *     #[GraphQLField]
 *     public string $name;
 * }
 * ```
 */
#[Attribute(Attribute::TARGET_CLASS)]
final readonly class GraphQLType
{
    /**
     * @param string|null $name Optional custom GraphQL type name. Defaults to class name.
     * @param string|null $description Optional description for schema documentation.
     * @param bool $isInput Whether this type represents a GraphQL input type.
     */
    public function __construct(
        public ?string $name = null,
        public ?string $description = null,
        public bool $isInput = false,
    ) {
    }
}

File: fraiseql-java/src/main/java/com/fraiseql/core/GraphQLType.java
package com.fraiseql.core;

import java.lang.annotation.*;

/**
 * Marks a class as a GraphQL type for FraiseQL schema authoring.
 *
 * Usage:
 * <pre>
 * @GraphQLType
 * public class User {
 *     @GraphQLField
 *     public int id;
 *
 *     @GraphQLField
 *     public String name;
 * }
 * </pre>
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface GraphQLType {
    /**
     * Optional custom name for the GraphQL type.
     * If not specified, the class name is used.
     */
    String name() default "";

    /**
     * Optional description for the GraphQL type.
     */
    String description() default "";

    /**
     * Whether this type implements the Relay Node interface.
     * When true, the type participates in global object identification via node(id: ID!)
     * and can be used as the return type of relay-enabled list queries.
     * Requires pk_{entity} (BIGINT) to be present in the view's data JSONB.
     */
    boolean relay() default false;
}
